File: components/maps/src/main/java/org/storm/maps/layer/TileLayer.java
package org.storm.maps.layer;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.image.Image;
import org.storm.core.ui.Resolution;
import org.storm.maps.tile.Tile;
import org.storm.maps.tile.TileSet;
import org.storm.physics.math.geometry.shapes.AxisAlignedRectangle;

import java.util.function.BiConsumer;

public class TileLayer extends Layer {

  private final TileSet tileSet;

  private final int[][] skeleton;

  public TileLayer(boolean active, Resolution resolution, TileSet tileSet, int[][] skeleton) {
    super(active, resolution);
    this.tileSet = tileSet;
    this.skeleton = skeleton;

    if (active) {
      this.loadTiles();
    }
  }

  private void loadTiles() {
    skeletonForEach(this.skeleton, (r, c) -> {
      if (this.skeleton[r][c] < 0) return;

      this.addEntity(new Tile(c * (double)this.tileSet.getTileWidth(), r * (double)this.tileSet.getTileHeight(),
        this.tileSet.getTileWidth(), this.tileSet.getTileHeight()));
    });
  }

  @Override
  public void render(GraphicsContext graphicsContext, double x, double y) {
    skeletonForEach(this.skeleton, (r, c) -> {
      if (this.skeleton[r][c] < 0) return;

      Image tileImage = this.tileSet.getTileImage(this.skeleton[r][c]);

      double tx = c * tileImage.getWidth();
      double ty = r * tileImage.getHeight();

      AxisAlignedRectangle tileRect = new AxisAlignedRectangle(tx, ty, tileImage.getWidth(), tileImage.getHeight());
      AxisAlignedRectangle screenRect = new AxisAlignedRectangle(x, y, this.resolution.getWidth(), this.resolution.getHeight());

      if (tileRect.intersects(screenRect)) {
        graphicsContext.drawImage(tileImage, tx - x, ty - y);
      }
    });
  }

  private static void skeletonForEach(int[][] skeleton, BiConsumer<Integer, Integer> skeletonConsumer) {
    for (int r = 0; r < skeleton.length; r++) {
      for (int c = 0; c < skeleton[0].length; c++) {
        skeletonConsumer.accept(r, c);
      }
    }
  }

  @Override
  public void update(double time, double elapsedTime) {
    // Tile Layers do not update by default
  }
}

File: components/maps/src/main/kotlin/org/storm/maps/layer/TileLayer.kt
package org.storm.maps.layer

import kotlinx.coroutines.runBlocking
import org.storm.core.context.Context
import org.storm.core.context.RESOLUTION
import org.storm.core.render.canvas.Canvas
import org.storm.maps.tile.Tile
import org.storm.maps.tile.TileSet
import org.storm.physics.math.geometry.shapes.AABB

class TileLayer(
    active: Boolean,
    private val tileSet: TileSet,
    private val skeleton: Array<IntArray>
) : Layer(active) {

    companion object {
        private suspend fun Array<IntArray>.iterateMatrix(block: suspend (Int, Int) -> Unit) {
            for (r in this.indices) {
                for (c in this[0].indices) {
                    block(r, c)
                }
            }
        }
    }

    init {
        if (active) {
            runBlocking { loadTiles() }
        }
    }

    override suspend fun render(canvas: Canvas, x: Double, y: Double) {
        val resolution = Context.RESOLUTION
        val screenRect = AABB(x, y, resolution.width, resolution.height)

        skeleton.iterateMatrix { r, c ->
            if (skeleton[r][c] < 0) return@iterateMatrix

            val tileImage = tileSet.tile(skeleton[r][c])
            val tx = c * tileImage.width
            val ty = r * tileImage.height

            val tileRect = AABB(tx, ty, tileImage.width, tileImage.height)

            // Only draw if its visible on the screen
            if (tileRect.intersects(screenRect)) {
                canvas.drawImageWithPixels(tileImage, tx - x, ty - y)
            }
        }
    }

    override suspend fun update(time: Double, elapsedTime: Double) {
        // Tile Layers do not update by default
    }

    private suspend fun loadTiles() {
        skeleton.iterateMatrix { r, c ->
            if (skeleton[r][c] < 0) return@iterateMatrix

            addEntity(
                Tile(
                    c * tileSet.tileWidth.toDouble(), r * tileSet.tileHeight.toDouble(),
                    tileSet.tileWidth.toDouble(), tileSet.tileHeight.toDouble()
                )
            )
        }
    }
}
